plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/alg/AbstractRandomListAccessor.java

```diff
@@ -0,0 +1,111 @@
+/*******************************************************************************
+ * Copyright (c) 2017 cds and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    cds - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.elk.core.alg;
+
+import java.util.List;
+
+import com.google.common.collect.Lists;
+
+/**
+ * Base class for classes that want to randomly access a list of elements. Access to non-existent list elements is
+ * explicitly permitted. The list is automatically enlarged by adding enough default values for the requested elements
+ * to exist. Subclasses must implement {@link #provideDefault()} to provide these default values.
+ * 
+ * @param <T> type of elements in the list.
+ */
+abstract class AbstractRandomListAccessor<T> {
```

Comment:
I'm not sure if the word  `random` is a good choice here. The access to the list is not really random, it's just that one doesn't know if the index was already set. 
Why not simply extend `ArrayList`, add the `getDefault` method and overwrite the `get` and `set` methods?